iml-services/iml-snapshot/src/retention.rs

```diff
@@ -0,0 +1,215 @@
+    .await?;
+
+    Ok(xs)
+}
+
+async fn get_retentions(pool: &sqlx::PgPool) -> Result<Vec<snapshot::SnapshotRetention>, Error> {
+    let xs = sqlx::query_as!(
+        snapshot::SnapshotRetention,
+        r#"
+                SELECT
+                    id,
+                    filesystem_name,
+                    reserve_value,
+                    reserve_unit as "reserve_unit:snapshot::ReserveUnit",
+                    last_run,
+                    keep_num
+                FROM snapshot_retention
+            "#
+    )
+    .fetch_all(pool)
+    .await?;
+
+    Ok(xs)
+}
+
+async fn get_retention_policy(
+    pool: &PgPool,
+    fs_name: &str,
+) -> Result<Option<snapshot::SnapshotRetention>, Error> {
+    let xs = get_retentions(pool).await?;
+
+    Ok(xs.into_iter().find(|x| x.filesystem_name == fs_name))
+}
+
+async fn destroy_snapshot(
+    client: Client,
+    fs_name: &str,
+    snapshot_name: &str,
+) -> Result<Command, Error> {
+    let resp: iml_graphql_queries::Response<iml_graphql_queries::snapshot::destroy::Resp> =
+        graphql(
+            client,
+            iml_graphql_queries::snapshot::destroy::build(fs_name, snapshot_name, true),
+        )
+        .await?;
+
+    let cmd = Result::from(resp)?.data.destroy_snapshot;
+
+    Ok(cmd)
+}
+
+async fn get_retention_filesystems(pool: &PgPool) -> Result<Vec<String>, Error> {
+    let xs = get_retentions(pool).await?;
+
+    let fs_names = xs.into_iter().map(|x| x.filesystem_name).collect();
+    Ok(fs_names)
+}
+
+pub async fn process_retention(
+    client: &Client,
+    influx_client: &InfluxClient,
+    pool: &PgPool,
+    mut stats_record: HashMap<String, u64>,
+) -> Result<HashMap<String, u64>, Error> {
+    let filesystems = get_retention_filesystems(pool).await?;
+
+    tracing::error!("Filesystems with retentions: {:?}", filesystems);
+
+    for fs_name in filesystems {
+        let stats = get_stats_from_influx(&fs_name, &influx_client).await?;
+
+        if let Some((bytes_total, bytes_free, bytes_used)) = stats {
+            tracing::debug!(
+                "stats values: {}, {}, {}",
+                bytes_total,
+                bytes_free,
+                bytes_used
+            );
+            let percent_used = bytes_used as f64 / bytes_total as f64 * 100.0f64;
+            let percent_free = 100.0f64 - percent_used;
+
+            tracing::error!(
+                "stats record: {:?} - bytes free: {}",
+                stats_record.get(&fs_name),
+                bytes_free
+            );
+            let retention = get_retention_policy(pool, &fs_name).await?;
+
+            if let Some(retention) = retention {
+                let snapshots = get_snapshots(pool, &fs_name).await?;
+
+                tracing::error!(
+                    "percent_left: {}, reserve value: {}",
+                    percent_free,
+                    retention.reserve_value
+                );
+                let delete_snapshot = match retention.reserve_unit {
```

Comment:
This var is named so like a decision is already made, but it is not.

Reply:
How does `should_delete_snapshot` sound? I think it makes sense when reading the if statement, "If should delete snapshot..."